Add a _RemoteDevice structure for ShardedTensor/ShardingSpec.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62927

As part of the ShardedTensor work, we realized we do need some sort of
_RemoteDevice structure that deals with our format of "workername/device" so
that users don't have to worry about parsing this string directly.

Right now this structure is just the bare minimum and is mostly a container for
describing a remote device. It is currently only used in ShardedTensor,
ShardingSpec and RemoteModule.

Once we actually have a consolidated remote device proposal, this class can be
extended appropriately if needed.

Differential Revision: [D30170689](https://our.internmc.facebook.com/intern/diff/D30170689/)

pritamdamania committed:
Add a _RemoteDevice structure for ShardedTensor/ShardingSpec.

As part of the ShardedTensor work, we realized we do need some sort of
_RemoteDevice structure that deals with our format of "workername/device" so
that users don't have to worry about parsing this string directly.

Right now this structure is just the bare minimum and is mostly a container for
describing a remote device. It is currently only used in ShardedTensor,
ShardingSpec and RemoteModule.

Once we actually have a consolidated remote device proposal, this class can be
extended appropriately if needed.

Differential Revision: [D30170689](https://our.internmc.facebook.com/intern/diff/D30170689/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Add a _RemoteDevice structure for ShardedTensor/ShardingSpec."


As part of the ShardedTensor work, we realized we do need some sort of
_RemoteDevice structure that deals with our format of "workername/device" so
that users don't have to worry about parsing this string directly.

Right now this structure is just the bare minimum and is mostly a container for
describing a remote device. It is currently only used in ShardedTensor,
ShardingSpec and RemoteModule.

Once we actually have a consolidated remote device proposal, this class can be
extended appropriately if needed.

Differential Revision: [D30170689](https://our.internmc.facebook.com/intern/diff/D30170689/)

[ghstack-poisoned]

wanchaol commented on torch/distributed/remote_device.py:
```diff
@@ -0,0 +1,111 @@
+                    and "cuda:1", just represent local devices.
+    """
+
+    def __init__(self, remote_device: Union[str, int]):
```
what does a `int` mean from user prospective? is it an index to local cuda device? I think we might want to accept a string or a torch.Device instead of int?

pritamdamania87 replied:
integer is similar to torch.device, where someone can do something like `torch.device(0)` and it is cuda:0. So you can do _RemoteDevice(0) and it will actually be just local cuda:0. Although, I do think your suggestion is better where it is only str or torch.device. I'll update this.

wanchaol commented on torch/distributed/_sharding_spec/api.py:
```diff
@@ -64,36 +59,26 @@ class ChunkShardingSpec(ShardingSpec):
 
     dim: ShardingDim
-    placements: List[Device]
+    placements: List[_RemoteDevice]
```
this seems not matching the doc, it should be `List[Union[_RemoteDevice, str]]`?

wanchaol commented on torch/distributed/_sharding_spec/_internals.py:
```diff
@@ -40,25 +16,19 @@ class ShardMetadata(object):
     def __post_init__(self):
-        if not is_valid_device(self.placement):
-            raise ValueError(f'{self.placement} is not a valid device')
+        if isinstance(self.placement, str):
```
so it seems `ShardedMetadata` still accepts a str and we convert to a `_RemoveDevice` if it's the case?

pritamdamania87 replied:
Yup, so you can pass in str or RemoteDevice for easy of use.

pritamdamania committed:
Update on "Add a _RemoteDevice structure for ShardedTensor/ShardingSpec."


As part of the ShardedTensor work, we realized we do need some sort of
_RemoteDevice structure that deals with our format of "workername/device" so
that users don't have to worry about parsing this string directly.

Right now this structure is just the bare minimum and is mostly a container for
describing a remote device. It is currently only used in ShardedTensor,
ShardingSpec and RemoteModule.

Once we actually have a consolidated remote device proposal, this class can be
extended appropriately if needed.

Differential Revision: [D30170689](https://our.internmc.facebook.com/intern/diff/D30170689/)

[ghstack-poisoned]

SciPioneer commented on torch/distributed/remote_device.py:
```diff
@@ -0,0 +1,111 @@
+import torch
+
+
+class _RemoteDevice(object):
```
Why not `@dataclass` this time?

mrshenli commented on torch/distributed/_sharding_spec/api.py:
```diff
@@ -1,11 +1,10 @@
 from abc import ABC
 from dataclasses import dataclass
 from typing import List, Union
+from torch.distributed.remote_device import _RemoteDevice
```
It looks good to me to have a dedicated type wrapping logic. 

(can be done in followup PRs) Regarding the naming, will it work if we follow existing `torch.device` 's convention? Sth like:

```python
remote_device = torch.distributed.device("worker0/cuda:0")
print(f"{remote_device.worker_name}, {remote_device.rank}, {remote_device.index}")
```

mrshenli replied:
I don't have a strong opinion on this. My feeling is that `distributed` might already be sufficient to tell it's not a local device. Please feel free to go with either one. :)

pritamdamania committed:
Update on "Add a _RemoteDevice structure for ShardedTensor/ShardingSpec."


As part of the ShardedTensor work, we realized we do need some sort of
_RemoteDevice structure that deals with our format of "workername/device" so
that users don't have to worry about parsing this string directly.

Right now this structure is just the bare minimum and is mostly a container for
describing a remote device. It is currently only used in ShardedTensor,
ShardingSpec and RemoteModule.

Once we actually have a consolidated remote device proposal, this class can be
extended appropriately if needed.

Differential Revision: [D30170689](https://our.internmc.facebook.com/intern/diff/D30170689/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Add a _RemoteDevice structure for ShardedTensor/ShardingSpec."


As part of the ShardedTensor work, we realized we do need some sort of
_RemoteDevice structure that deals with our format of "workername/device" so
that users don't have to worry about parsing this string directly.

Right now this structure is just the bare minimum and is mostly a container for
describing a remote device. It is currently only used in ShardedTensor,
ShardingSpec and RemoteModule.

Once we actually have a consolidated remote device proposal, this class can be
extended appropriately if needed.

Differential Revision: [D30170689](https://our.internmc.facebook.com/intern/diff/D30170689/)

[ghstack-poisoned]
